### PR TITLE
Add support for wait_readable() and wait_writable() on ProactorEventLoop

### PR DESCRIPTION
**Pull request by agronholm**
## Changes

This enables asyncio on Windows to call `wait_readable()` and `wait_writable()` by using a selector on a separate thread. 

Closes #820.

## Checklist

If this is a user-facing code change, like a bugfix or a new feature, please ensure that
you've fulfilled the following conditions (where applicable):

- [X] You've added tests (in `tests/`) added which would fail without your patch
- [X] You've updated the documentation (in `docs/`, in case of behavior changes or new
features)
- [X] You've added a new changelog entry (in `docs/versionhistory.rst`).

If this is a trivial change, like a typo fix or a code reformatting, then you can ignore
these instructions.

### Updating the changelog

If there are no entries after the last release, use `**UNRELEASED**` as the version.
If, say, your patch fixes issue #123, the entry should look like this:

`* Fix big bad boo-boo in task groups (#123
<https://github.com/agronholm/anyio/issues/123>_; PR by @yourgithubaccount)`

If there's no issue linked, just link to your pull request instead by updating the
changelog after you've created the PR.

### CODE REVIEW

**Review by davidbrochart — APPROVED** [review]

File: src/anyio/_backends/_asyncio.py [hunk 1]
```diff
@@ -2776,22 +2782,26 @@ async def wait_writable(cls, obj: HasFileno | int) -> None:
...
-            raise BusyResourceError("writing to") from None
+            raise BusyResourceError("writing to")
 
         loop = get_running_loop()
-        event = write_events[obj] = asyncio.Event()
-        loop.add_writer(obj, event.set)
+        event = asyncio.Event()
+        try:
+            loop.add_writer(obj, event.set)
+            remove_writer = loop.remove_writer
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:2791]: Nit: I think this should be in the except's else clause

File: src/anyio/_core/_asyncio_selector_thread.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+_selector_lock = threading.Lock()
+_selector: Selector | None = None
+
+
+class Selector:
+    def __init__(self) -> None:
+        self._thread = threading.Thread(target=self.run, name="AnyIO socket selector")
+        self._selector = DefaultSelector()
+        self._send, self._receive = socket.socketpair()
```
Comment (graingert) [inline src/anyio/_core/_asyncio_selector_thread.py:21]: I think both sockets should be non-blocking, and you should ignore BlockingIOError on the send side

File: src/anyio/_core/_asyncio_selector_thread.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        else:
+            self._selector.unregister(fd)
+
+        return True
+
+    def run(self) -> None:
+        while not self._closed:
+            for key, events in self._selector.select():
+                if key.fileobj is self._receive:
+                    self._receive.recv(10240)
```
Comment (graingert) [inline src/anyio/_core/_asyncio_selector_thread.py:106]: I think this needs to be a loop with a smaller buffer size:

https://github.com/python/cpython/blob/3.13/Lib/asyncio/selector_events.py#L130

File: src/anyio/_core/_asyncio_selector_thread.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                if events & EVENT_WRITE:
+                    loop, callback = key.data[EVENT_WRITE]
+                    self.remove_writer(key.fd)
+                    try:
+                        loop.call_soon_threadsafe(callback)
+                    except RuntimeError:
+                        pass  # the loop was already closed
+
+        self._selector.unregister(self._receive)
+        self._receive.close()
```
Comment (graingert) [inline src/anyio/_core/_asyncio_selector_thread.py:126]: I think this should be closed using `with`

File: src/anyio/_core/_sockets.py [hunk 5]
```diff
@@ -679,25 +673,22 @@ def wait_readable(obj: HasFileno | int) -> Awaitable[None]:
     return get_async_backend().wait_readable(obj)
 
 
-def wait_writable(obj: HasFileno | int) -> Awaitable[None]:
+def wait_writable(obj: FileDescriptorLike) -> Awaitable[None]:
     """
     Wait until the given object can be written to.
 
-    This does **NOT** work on Windows when using the asyncio backend with a proactor
```
Comment (graingert) [inline src/anyio/_core/_sockets.py:686]: Do you think it's worth mentioning the thread implementation details here?

**Review by graingert — CHANGES_REQUESTED** [review]
Looks pretty great overall, this is going to be a killer feature. 

Just some nits, and I think you need to tweak the socketpair impl

File: src/anyio/_backends/_asyncio.py [hunk 6]
```diff
@@ -2746,25 +2748,29 @@ async def wait_readable(cls, obj: HasFileno | int) -> None:
...
-            raise BusyResourceError("reading from") from None
+            raise BusyResourceError("reading from")
 
         loop = get_running_loop()
-        event = read_events[obj] = asyncio.Event()
-        loop.add_reader(obj, event.set)
+        event = asyncio.Event()
+        try:
+            loop.add_reader(obj, event.set)
+            remove_reader = loop.remove_reader
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:2757]: Nit: I think this should be in the except's else clause

File: src/anyio/_core/_asyncio_selector_thread.py (hunk 3 above)
Comment (graingert) [reply src/anyio/_core/_asyncio_selector_thread.py:106]: Possibly both of these should be recv_into?

File: src/anyio/_core/_asyncio_selector_thread.py (hunk 4 above)
Comment (graingert) [reply src/anyio/_core/_asyncio_selector_thread.py:126]: I also think you need to close the selector https://docs.python.org/3/library/selectors.html#selectors.BaseSelector.close

File: src/anyio/_core/_asyncio_selector_thread.py (hunk 3 above)
Comment (agronholm) [reply src/anyio/_core/_asyncio_selector_thread.py]: I get `BlockingIOError`, but why `InterruptedError` too? System calls are automatically retried by Python whenever they're interrupted.

File: src/anyio/_core/_asyncio_selector_thread.py (hunk 2 above)
Comment (agronholm) [reply src/anyio/_core/_asyncio_selector_thread.py:21]: Done.

File: src/anyio/_backends/_asyncio.py (hunk 1 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:2791]: I thought someone might pick up on that...

File: src/anyio/_backends/_asyncio.py (hunk 6 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Fixed.

File: src/anyio/_backends/_asyncio.py (hunk 1 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Fixed.

File: src/anyio/_core/_asyncio_selector_thread.py (hunk 4 above)
Comment (agronholm) [reply src/anyio/_core/_asyncio_selector_thread.py]: I'm closing the selector and the socket in `stop()` now.

**Comment by agronholm** [comment]
I messed up somewhere along the way and now it gets stuck in a busy-wait loop somewhere.

File: src/anyio/_backends/_asyncio.py [hunk 7]
```diff
@@ -2788,13 +2789,14 @@ async def wait_writable(cls, obj: FileDescriptorLike) -> None:
         event = asyncio.Event()
         try:
             loop.add_writer(obj, event.set)
-            remove_writer = loop.remove_writer
         except NotImplementedError:
             from anyio._core._asyncio_selector_thread import get_selector
 
             selector = get_selector()
             selector.add_writer(obj, event.set)
             remove_writer = selector.remove_writer
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:2797]: Looks like this line should be deleted

**Review by graingert — CHANGES_REQUESTED** [review]

File: src/anyio/_core/_asyncio_selector_thread.py [hunk 8]
```diff
@@ -114,7 +114,12 @@ def run(self) -> None:
         while not self._closed:
             for key, events in self._selector.select():
                 if key.fileobj is self._receive:
-                    self._receive.recv(4096)
+                    while True:
+                        try:
+                            self._receive.recv(4096)
```
Comment (graingert) [inline src/anyio/_core/_asyncio_selector_thread.py:119]: You need to make the socket non blocking

Comment (agronholm) [reply src/anyio/_core/_asyncio_selector_thread.py:119]: But I did! Lines 22 and 23.

File: src/anyio/_core/_sockets.py (hunk 5 above)
Comment (agronholm) [reply src/anyio/_core/_sockets.py:686]: Done.

File: src/anyio/_core/_sockets.py [hunk 9]
```diff
@@ -663,10 +660,12 @@ def wait_readable(obj: HasFileno | int) -> Awaitable[None]:
     descriptors aren't supported, and neither are handles that refer to anything besides
     a ``SOCKET``.
 
-    This does **NOT** work on Windows when using the asyncio backend with a proactor
-    event loop (default on py3.8+).
+    On backends where this functionality is not natively provided (asyncio
+    ``ProactorEventLoop`` on Windows) Additionally, on asyncio, this functionality is
```
Comment (graingert) [inline src/anyio/_core/_sockets.py:664]:
```suggestion
    ``ProactorEventLoop`` on Windows), this functionality is
```

Comment (agronholm) [reply src/anyio/_core/_sockets.py]: Woops. I pushed a better wording just now.

**Review by graingert — APPROVED** [review]

**Pull request merged**